### ISSUE TITLE
std.crypto: curve448 support

### ISSUE THREAD

**Issue by eightfilms** [ziglang/zig#22101]
paper here: https://eprint.iacr.org/2015/625.pdf

I see that there is curve25519 support, but not curve448. Wondering if it would be worth adding to `std.crypto`, or if there are plans to do so.

cc: @jedisct1

**Comment by jedisct1** [comment]
Hi!

Curve448 has very little adoption besides Tor.

Nowadays, the threat is quantum computers, that would break Curve448 as quickly as Curve25519. And since 448 is slower while providing no practical security benefits, it's a bit difficult to justify.

That being said, it's easy to implement on the basis on the work already done for 25519, so it will be eventually added. But it's probably not super high priority.

**Comment by eightfilms** [comment]
Sounds reasonable, thanks for the response!

**Comment by aidenfoxivey** [comment]
> Hi!
> 
> Curve448 has very little adoption besides Tor.
> 
> Nowadays, the threat is quantum computers, that would break Curve448 as quickly as Curve25519. And since 448 is slower while providing no practical security benefits, it's a bit difficult to justify.
> 
> That being said, it's easy to implement on the basis on the work already done for 25519, so it will be eventually added. But it's probably not super high priority.

I'd be happy to write a PR to add it. Would that be looked upon favourably?

**Comment by jedisct1** [comment]
> I'd be happy to write a PR to add it. Would that be looked upon favourably?

In order to get parity and consistency with Curve25519, this is going to be a lot of code (EC and scalar arithmetic, decaf, ecdh, signatures, TLS support), so a single ginormous PR wouldn't be ideal.

I'd suggest working on this in a dedicated repository that can be added as a dependency, and once it's complete and tested, it will be easier to merge.

I'm still a bit worried about this in the standard library, though. Wondering if this is actually going to be used, or of this is code bloat that very few applications are actually going to use, but that we'll have to maintain forever.

Besides Tor, what is actually requiring Curve448? Even in the context of TLS, are there servers that require Curve448 and don't support Curve25519 at all?

**Comment by eightfilms** [comment]
> Besides Tor, what is actually requiring Curve448? Even in the context of TLS, are there servers that require Curve448 and don't support Curve25519 at all?

To add some context to my original question, I was just looking at implementing [noise protocol](https://noiseprotocol.org/noise.html) in zig and one of the two DH functions uses curve448, so it's not out of any current practical/production need - my original question was just out of curiosity.

**Comment by aidenfoxivey** [comment]
> > I'd be happy to write a PR to add it. Would that be looked upon favourably?
> 
> In order to get parity and consistency with Curve25519, this is going to be a lot of code (EC and scalar arithmetic, decaf, ecdh, signatures, TLS support), so a single ginormous PR wouldn't be ideal.
> 
> I'd suggest working on this in a dedicated repository that can be added as a dependency, and once it's complete and tested, it will be easier to merge.
> 
> I'm still a bit worried about this in the standard library, though. Wondering if this is actually going to be used, or of this is code bloat that very few applications are actually going to use, but that we'll have to maintain forever.
> 
> Besides Tor, what is actually requiring Curve448? Even in the context of TLS, are there servers that require Curve448 and don't support Curve25519 at all?

Fair enough! If there's sufficient interest, I would consider working on it. :)

**Comment by sorairolake** [comment]
If we add this elliptic curve, I think it would be better to add the following algorithms as well:

- X448 ([RFC 7748](https://datatracker.ietf.org/doc/html/rfc7748))
- Ed448 ([RFC 8032](https://datatracker.ietf.org/doc/html/rfc8032))
- decaf448 ([RFC 9496](https://datatracker.ietf.org/doc/html/rfc9496))

According to [RFC 8731](https://datatracker.ietf.org/doc/html/rfc8731), [RFC 8709](https://datatracker.ietf.org/doc/html/rfc8709) and [RFC 9580](https://datatracker.ietf.org/doc/html/rfc9580), SSH and OpenPGP supports *448. So adding *448 may be useful when implementing these or TLS.

I think most protocols which supports *448 also supports *25519. I think it is still rare for these protocols to use *448 instead of *25519. So I don't think there is enough demand yet to add this elliptic curve and the algorithms built on top of it to the standard library.

**Comment by jedisct1** [comment]
This is probably not going to happen in the standard library.
